// @expressots/core/lib/esm/environment/env-validator.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var EnvValidatorProvider_1;
import fs from "fs";
import path from "path";
import dotenv from "dotenv";
import { provide } from "inversify-binding-decorators";
import { LogLevel, log } from "../logger";
/**
 * The EnvValidatorProvider class provides utility methods for working with environment variables.
 * It validates, loads, and retrieves environment variables from the .env file.
 * @provide EnvValidatorProvider
 */
let EnvValidatorProvider = EnvValidatorProvider_1 = class EnvValidatorProvider {
    /**
     * Retrieves the value of an environment variable, or a default value if the variable is not set.
     * @param key - The key of the environment variable.
     * @param defaultValue - The default value to return if the environment variable is not set.
     * @returns The value of the environment variable, or the default value if not set.
     */
    static get(key, defaultValue = undefined) {
        return process.env[key] ?? defaultValue;
    }
    /**
     * Validates and loads all environment variables from the .env file.
     * If the .env file does not exist or any environment variables are not set, the process will exit with an error.
     */
    static checkAll() {
        /* Load .env file */
        dotenv.config();
        /* Verify if .env file exists */
        const envFilePath = path.join(process.cwd(), ".", ".env");
        if (!fs.existsSync(envFilePath)) {
            log(LogLevel.Info, "Environment file .env is not defined.", "env-validator-provider");
            process.exit(1);
        }
        const regexIgnoreDefaultEnvKeys = /^npm_config_/;
        let hasError = false;
        for (const key in process.env) {
            if (regexIgnoreDefaultEnvKeys.test(key)) {
                continue;
            }
            if (!process.env[key] || process.env[key] === "") {
                log(LogLevel.Info, `Environment variable ${key} is not defined.`, "env-validator-provider");
                hasError = true;
            }
        }
        if (hasError) {
            process.exit(1);
        }
    }
};
EnvValidatorProvider = EnvValidatorProvider_1 = __decorate([
    provide(EnvValidatorProvider_1)
], EnvValidatorProvider);
String.prototype.AsBoolean = function () {
    switch (this.toLowerCase().trim()) {
        case "true":
        case "1":
        case "yes":
            return true;
        case "false":
        case "0":
        case "no":
            return false;
        default:
            return undefined;
    }
};
String.prototype.AsNumber = function () {
    return Number(this);
};
String.prototype.AsString = function () {
    return String(this);
};
export { EnvValidatorProvider as Environments };
