// @expressots/core/lib/esm/application/application.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var Application_1;
import express from "express";
import { provide } from "inversify-binding-decorators";
import { InversifyExpressServer } from "inversify-express-utils";
import process from "process";
import { Console } from "../console/console";
import errorHandler from "../error/error-handler-middleware";
/**
 * Enum representing possible server environments.
 */
var ServerEnvironment;
(function (ServerEnvironment) {
    ServerEnvironment["Development"] = "development";
    ServerEnvironment["Staging"] = "staging";
    ServerEnvironment["Production"] = "production";
})(ServerEnvironment || (ServerEnvironment = {}));
/**
 * The Application class provides a way to configure and manage an Express application.
 * @provide Application
 */
let Application = Application_1 = class Application {
    app;
    port;
    environment;
    /**
     * Constructs a new instance of the Application class.
     */
    constructor() { }
    /**
     * Configure services that should be initialized before the server starts.
     */
    configureServices() { }
    /**
     * Configure services that should be executed after the server starts.
     */
    postServerInitialization() { }
    /**
     * Perform actions or cleanup after the server is shutdown.
     */
    serverShutdown() {
        process.exit(0);
    }
    /**
     * Create and configure the Express application.
     * @param container - The InversifyJS container.
     * @param middlewares - An array of Express middlewares to be applied.
     * @returns The configured Application instance.
     */
    create(container, middlewares = []) {
        this.configureServices();
        const expressServer = new InversifyExpressServer(container);
        expressServer.setConfig((app) => {
            /* Default body parser application/json */
            app.use(express.json());
            /* Default body parser application/x-www-form-urlencoded */
            app.use(express.urlencoded({ extended: true }));
            middlewares.forEach(middleware => {
                app.use(middleware);
            });
        });
        this.app = expressServer.build();
        /* Add the error handler middleware */
        this.app.use(errorHandler);
        return this;
    }
    /**
     * Start listening on the given port and environment.
     * @param port - The port number to listen on.
     * @param environment - The server environment.
     * @param consoleMessage - Optional message to display in the console.
     */
    listen(port, environment, consoleMessage) {
        this.port = port;
        this.environment = environment;
        this.app.listen(this.port, () => {
            new Console().messageServer(this.port, this.environment, consoleMessage);
            process.on("SIGINT", this.serverShutdown.bind(this));
        });
        this.postServerInitialization();
    }
};
Application = Application_1 = __decorate([
    provide(Application_1),
    __metadata("design:paramtypes", [])
], Application);
const appServerInstance = new Application();
export { appServerInstance as AppInstance, Application, ServerEnvironment };
